diff --git a/src/rtklib.h b/src/rtklib.h
index 3f3287a..8510f3f 100644
--- a/src/rtklib.h
+++ b/src/rtklib.h
@@ -432,6 +432,7 @@ extern "C" {
 #define STR_UDPSVR   12                 /* stream type: UDP server */
 #define STR_UDPCLI   13                 /* stream type: UDP server */
 #define STR_MEMBUF   14                 /* stream type: memory buffer */
+#define STR_UNIXSVR  15                 /* stream type: unix domain socket server */
 
 #define STRFMT_RTCM2 0                  /* stream format: RTCM 2 */
 #define STRFMT_RTCM3 1                  /* stream format: RTCM 3 */
